osu.Framework.Tests/Graphics/LifetimeEntryManagerTest.cs

```diff
@@ -0,0 +1,342 @@
+
+            manager.AddEntry(entry);
+            checkCountAliveAt(0, 1);
+
+            manager.RemoveEntry(entry);
+            checkCountAliveAt(0, 0);
+
+            manager.AddEntry(entry);
+            checkCountAliveAt(0, 1);
+        }
+
+        [Test]
+        public void TestDynamicChange()
+        {
+            manager.AddEntry(new LifetimeEntry { LifetimeStart = -1, LifetimeEnd = 0 });
+            manager.AddEntry(new LifetimeEntry { LifetimeStart = 0, LifetimeEnd = 1 });
+            manager.AddEntry(new LifetimeEntry { LifetimeStart = 0, LifetimeEnd = 1 });
+            manager.AddEntry(new LifetimeEntry { LifetimeStart = 1, LifetimeEnd = 2 });
+
+            checkCountAliveAt(0, 2);
+
+            manager.Entries[0].LifetimeEnd = 1;
+            manager.Entries[1].LifetimeStart = 1;
+            manager.Entries[2].LifetimeEnd = 0;
+            manager.Entries[3].LifetimeStart = 0;
+
+            checkCountAliveAt(0, 2);
+
+            foreach (var entry in manager.Entries)
+            {
+                entry.LifetimeEnd += 1;
+                entry.LifetimeStart += 1;
+            }
+
+            checkCountAliveAt(0, 1);
+
+            foreach (var entry in manager.Entries)
+            {
+                entry.LifetimeStart -= 1;
+                entry.LifetimeEnd -= 1;
+            }
+
+            checkCountAliveAt(0, 2);
+        }
+
+        [Test]
+        public void TestBoundaryCrossing()
+        {
+            manager.AddEntry(new LifetimeEntry { LifetimeStart = -1, LifetimeEnd = 0 });
+            manager.AddEntry(new LifetimeEntry { LifetimeStart = 0, LifetimeEnd = 1 });
+            manager.AddEntry(new LifetimeEntry { LifetimeStart = 1, LifetimeEnd = 2 });
+
+            // No crossings for the first update.
+            manager.Update(0);
+            checkNoCrossing(manager.Entries[0]);
+            checkNoCrossing(manager.Entries[1]);
+            checkNoCrossing(manager.Entries[2]);
+
+            manager.Update(2);
+            checkNoCrossing(manager.Entries[0]);
+            checkCrossing(manager.Entries[1], 0, LifetimeBoundaryKind.End, LifetimeBoundaryCrossingDirection.Forward);
+            checkCrossing(manager.Entries[2], 0, LifetimeBoundaryKind.Start, LifetimeBoundaryCrossingDirection.Forward);
+            checkCrossing(manager.Entries[2], 1, LifetimeBoundaryKind.End, LifetimeBoundaryCrossingDirection.Forward);
+
+            manager.Update(1);
+            checkNoCrossing(manager.Entries[0]);
+            checkNoCrossing(manager.Entries[1]);
+            checkCrossing(manager.Entries[2], 0, LifetimeBoundaryKind.End, LifetimeBoundaryCrossingDirection.Backward);
+
+            manager.Update(-1);
+            checkCrossing(manager.Entries[0], 0, LifetimeBoundaryKind.End, LifetimeBoundaryCrossingDirection.Backward);
+            checkCrossing(manager.Entries[1], 0, LifetimeBoundaryKind.End, LifetimeBoundaryCrossingDirection.Backward);
```

Comment:
For completeness, should also assert
```suggestion
            checkCrossing(manager.Entries[1], 0, LifetimeBoundaryKind.End, LifetimeBoundaryCrossingDirection.Backward);
            checkCrossing(manager.Entries[1], 1, LifetimeBoundaryKind.Start, LifetimeBoundaryCrossingDirection.Backward);
```